Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h"
+
+#import <Foundation/Foundation.h>
+#import <SystemConfiguration/SystemConfiguration.h>
+#import <netinet/in.h>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "absl/memory/memory.h"
+#include "dispatch/dispatch.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+namespace {
+
+using NetworkStatus = ConnectivityMonitor::NetworkStatus;
+using util::AsyncQueue;
+
+NetworkStatus ToNetworkStatus(SCNetworkReachabilityFlags flags) {
+  if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+    return NetworkStatus::Unreachable;
+  }
+  if (flags & kSCNetworkReachabilityFlagsConnectionRequired) {
+    return NetworkStatus::Unreachable;
+  }
+
+#if TARGET_OS_IPHONE
+  if (flags & kSCNetworkReachabilityFlagsIsWWAN) {
+    return NetworkStatus::ReachableViaCellular;
+  }
+#endif
+  return NetworkStatus::Reachable;
+}
+
+SCNetworkReachabilityRef CreateReachability() {
+  // Pseudoaddress that monitors internet reachability in general.
+  sockaddr_in any_connection_addr{};
+  any_connection_addr.sin_len = sizeof(any_connection_addr);
+  any_connection_addr.sin_family = AF_INET;
+  return SCNetworkReachabilityCreateWithAddress(
+      nullptr, reinterpret_cast<sockaddr*>(&any_connection_addr));
+}
+
+}  // namespace
+
+/**
+ * Implementation of `ConnectivityMonitor` based on `SCNetworkReachability`
+ * (iOS/MacOS).
+ */
+class ConnectivityMonitorApple : public ConnectivityMonitor {
+ public:
+  ConnectivityMonitorApple(AsyncQueue* worker_queue)
+      : ConnectivityMonitor{worker_queue}, reachability_{CreateReachability()} {
+    SCNetworkReachabilityFlags flags;
+    if (SCNetworkReachabilityGetFlags(reachability_, &flags)) {
```

Comment:
I'm pretty confused about whether this method can block. The documentation doesn't mention it, but it is unhelpful in other ways, too. The [usage example](https://developer.apple.com/library/archive/samplecode/Reachability/Listings/ReadMe_md.html#//apple_ref/doc/uid/DTS40007324-ReadMe_md-DontLinkElementID_11) mentions:
> _do not issue a synchronous check by hostName on the main thread. If the device cannot reach a DNS server or is on a slow network, a synchronous call to the SCNetworkReachabilityGetFlags function can block for up to 30 seconds trying to resolve the hostName. If this happens on the main thread, the application watchdog will kill the application after 20 seconds of inactivity._

This is pretty clear for the case an actual hostname is used, but I'm not sure whether this applies to that pseudoaddress that simply checks for general internet availability. FWIW, gRPC does this [right in the initializer](https://github.com/grpc/grpc/blob/master/src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m#L66-L68), but I'm not sure on which queue it gets called. Doing some experimenting, the function seems to return very quickly whether the network is up or down.